### PR TITLE
ci: daggerize test and lint pipelines

### PR DESCRIPTION
**Pull request by marcosnils**
:wave:  coming from #1436 

TL;DR: 

1. Install the `dagger` CLI (https://docs.dagger.io/quickstart/cli)
2. Check out this PR and run:

```
dagger --src . call make-all --args "check_license test"
```

you can alternatively tell dagger to run the tests against this PR's branch without having to check out your code in your machine as follows:

```
dagger call -m github.com/prometheus/client_golang@pull/1534/merge --src https://github.com/prometheus/client_golang#pull/1534/merge \
 make-all --args "check_license test"
```

This runs the same tests the current `GHA` pipeline runs here https://github.com/prometheus/client_golang/actions/runs/9419021570/workflow#L60 for all the `supported_go_versions.txt` file.

---

Hopefully this serves as a starting point to make this project's CI pipelines more reproducible and consistent across different environments :pray: .

Some considerations I took while daggerizing this pipeline: 

-  I've decided to "wrap" as much as I could in order to make the change the least disruptive. As you can see from the PR, I kept the same `make` UX the current repository has 
- I mostly focused on the `go.yml` (https://github.com/prometheus/client_golang/actions/runs/9419021570/workflow)  and `golangci-lint.yml` (https://github.com/prometheus/client_golang/actions/runs/9419021582/workflow)  workflows as those are mostly the ones that get executed for each PR and commit to `main`. 
- I've noticed that the GHA workflow files seem to updated from the `prometheus/prometheus`  repository. Given that I needed to modify them, I just did it in this repo as I'm unsure which would be the proper way to proceed here.

cc @kakkoyun

### CODE REVIEW

**Comment by shykes** [comment]
> you can alternatively tell dagger to run the tests against this PR's branch without having to check out your code in your machine as follows:
> 
> ```
> dagger -m github.com/prometheus/client_golang@refs/pull/1534/merge call --src git@github.com:prometheus/client_golang#refs/pull/1534/merge make-all --args "check_license test"
> ```

For some reason that git URL failed for me, but this worked:

```console
dagger call -m github.com/prometheus/client_golang@pull/1534/merge --src https://github.com/prometheus/client_golang#pull/1534/merge \
 make-all --args "check_license test"
```

I get some test failures, is that expected @marcosnils ?

```console
[...]
FAIL    github.com/prometheus/client_golang/prometheus  16.423s
ok      github.com/prometheus/client_golang/prometheus/collectors       0.022s
ok      github.com/prometheus/client_golang/prometheus/graphite 0.026s
ok      github.com/prometheus/client_golang/prometheus/internal 0.026s
ok      github.com/prometheus/client_golang/prometheus/promauto 0.026s
ok      github.com/prometheus/client_golang/prometheus/promhttp 0.386s
ok      github.com/prometheus/client_golang/prometheus/push     0.023s
ok      github.com/prometheus/client_golang/prometheus/testutil 0.026s
ok      github.com/prometheus/client_golang/prometheus/testutil/diff    0.023s
ok      github.com/prometheus/client_golang/prometheus/testutil/promlint        0.023s
FAIL
Stderr:
make: *** [Makefile.common:153: common-test] Error
```

**Comment by marcosnils** [comment]
> For some reason that git URL failed for me, but this worked:

ok, strange.. both seem to work for me. I'll update the original URL to use yours in the original description :pray:

**Comment by marcosnils** [comment]
> I get some test failures, is that expected @marcosnils ?

just :eyes: at this failures and I have the theory that they're related to some flukes due to timing nits in tests: https://github.com/marcosnils/client_golang/blob/e56689d0f2d0bbff8be277c2e4a8dd71c89ce47a/prometheus/summary_test.go?plain=1#L389.  Since `make-all` runs all go tests in parallel, looks like it's causing that tests to match incorrect assertions. 

cc @kakkoyun WDYT?

**Comment by kakkoyun** [comment]
@marcosnils Awesome! Thanks for the kickstart. I'll check this PR in-depth in the upcoming days. And see what we can improve.

**Comment by marcosnils** [comment]
hey @kakkoyun gentle ping here. LMK if you want me to rebase this PR if you have the time to :eyes: :pray:

File: dagger/main.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+		Container().
+		WithMountedDirectory("/src", m.Source).
+		WithWorkdir("/src").
+		WithMountedCache("/go/bin", dag.CacheVolume("gobincache")).
+		WithExec([]string{"sh", "-c", "make " + args}).
+		Stdout(context.Background())
+}
+
+// runs `make` with the given arguments for all supported go versions
+func (m *ClientGolang) MakeAll(
```
Comment (kakkoyun) [inline dagger/main.go:53]: I think `make all` is a bit excessive to run in this repo. We can live with
- `make test`
- `make fmt`
for starters.

This could also solve the issue with the license checker that you had to edit above.

Comment (kakkoyun) [reply dagger/main.go:53]: And maybe `make lint`

**Review by kakkoyun — COMMENTED** [review]
This is looking good. 🗡️ 🗡️ 🗡️ 

Sorry for the extremely late review, and you had to ping us a couple of times. 

This is super close to done. If you can address the comments we have put in here.

We review this together with @ArthurSens (hence the "we"s 😜)

**Comment by kakkoyun** [comment]
@marcosnils, @ArthurSens, and I finally had some time to spare on this. Thanks again for the contribution.

File: dagger/go.mod [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+require (
+	github.com/cenkalti/backoff/v4 v4.3.0 // indirect
+	github.com/go-logr/logr v1.4.1 // indirect
+	github.com/go-logr/stdr v1.2.2 // indirect
+	github.com/google/uuid v1.6.0 // indirect
+	github.com/grpc-ecosystem/grpc-gateway/v2 v2.19.1 // indirect
+	github.com/sosodev/duration v1.2.0 // indirect
+	go.opentelemetry.io/otel/exporters/otlp/otlptrace v1.26.0 // indirect
+	go.opentelemetry.io/otel/metric v1.26.0 // indirect
```
[the remark this answers is not in the document]
Comment (marcosnils) [reply dagger/go.mod:27]: :D that's a good observations. These imports are coming from upstream Dagger. I'd love to learn more what the difference would be by using `client_golang` instead of OTEL standard libs here.  cc @vito

**Comment by marcosnils** [comment]
@kakkoyun @ArthurSens I've addressed all the comments above. Please let me know if there's anything else that you'd prefer to be changed :pray:

File: dagger/main.go (hunk 1 above)
Comment (bwplotka) [reply dagger/main.go:54]: Let's call it `MakeLintTest` then?

File: .github/workflows/golangci-lint.yml [hunk 3]
```diff
@@ -26,7 +26,7 @@ jobs:
       - name: Checkout repository
         uses: actions/checkout@a5ac7e51b41094c92402da3b24376905380afc29 # v4.1.6
       - name: Install Go
-        uses: actions/setup-go@0a12ed9d6a96ab950c8f026ed9f722fe0da7ef32 # v5.0.2
+        uses: actions/setup-go@0c52d547c9bc32b1aa3301fd7a9cb496313a4491 # v5.0.0
```
Comment (bwplotka) [inline .github/workflows/golangci-lint.yml:29]: do we need this file with Dagger?

**Review by bwplotka — APPROVED** [review]
Thanks! I think it's bit adventurous, but given we have little CI scaffolding let's experiment and see how it goes 💪🏽  However, I would try to actively remove old paths to not have two ways of doing things (that means slowly migrating to dagger/main from Make right?

File: .github/workflows/golangci-lint.yml (hunk 3 above)
Comment (marcosnils) [reply .github/workflows/golangci-lint.yml]: no really but this file is being automatically synced by the prombot and @kakkoyun commented above that we might not want to modify it

**Comment by marcosnils** [comment]
> things (that means slowly migrating to dagger/main from Make right?

This would be the ideal. At the same time, since Dagger is still calling the same underlying make targets, IMO it's a "good enough" trade-off as an initial approach to see how Dagger feel :relaxed:

File: dagger/main.go (hunk 1 above)
Comment (kakkoyun) [reply dagger/main.go:54]: Ok, yes, the name is confusing, then. `MakeRun` or `MakeRunWithSupportedVersions` would make more sense to me.

**Review by kakkoyun — COMMENTED** [review]
LGTM.

If we can fix the naming and add a comment to explain the intention, we are golden. 

We can merge it after a rebase to make sure we capture all the sync files from upstream.

Thanks again @marcosnils

**Comment by marcosnils** [comment]
> If we can fix the naming and add a comment to explain the intention, we are golden.

done! the comment was already there :open_hands:  (https://github.com/prometheus/client_golang/pull/1534/files#diff-f18b612d33c74a552af8f75da09a36af695ef16d51ef7537bd1e7ef2d3d09685R53)

**Review by kakkoyun — APPROVED** [review]
